### PR TITLE
feat: Add download progress to sidebar

### PR DESCRIPTION
**Pull request by oto-ciulis-tt**
**Note:** Requires https://github.com/Comfy-Org/electron/pull/214 to work properly

https://github.com/user-attachments/assets/e70562a4-d7e7-4a65-8485-bc11cac8ccc6

image

### CODE REVIEW

**Review by huchenlei — CHANGES_REQUESTED** [review]
Sorry we probably didn't communicate this clear enough. The fault is probably on me and Robin. I think the requirement is simpler than the current impl. We think moving the list of download items (Progress bar + Action buttons) before the models tree should be good enough, i.e. we don't need to mess with the tree impl.

File: src/components/sidebar/tabs/ModelLibrarySidebarTab.vue [hunk 1]
```diff
@@ -183,10 +215,61 @@ watch(
...
+  data: {}
+})
+
 onMounted(async () => {
   if (settingStore.get('Comfy.ModelLibrary.AutoLoadAll')) {
     await modelStore.loadModels()
   }
+
+  if (isElectron()) {
+    const allDownloads: ModelDownload[] =
```
Comment (huchenlei) [inline src/components/sidebar/tabs/ModelLibrarySidebarTab.vue]: Let's store the download tasks in a Pinia store, so that it becomes a global state accessible by all components. It would also help us keep the component decoupled from the downloading logic.

**Comment by robinjhuang** [comment]
For cancelled / errored out items, can we add an option to remove the item from the list? Can add that in next PR.

File: src/types/treeExplorerTypes.ts [hunk 2]
```diff
@@ -43,6 +43,7 @@ export interface TreeExplorerNode<T = any> {
   contextMenuItems?:
     | MenuItem[]
     | ((targetNode: RenderedTreeExplorerNode) => MenuItem[])
+  progress?: number
```
Comment (huchenlei) [inline src/types/treeExplorerTypes.ts]: nit: revert unnecessary change here.

File: src/components/sidebar/tabs/modelLibrary/ModelTreeLeaf.vue [hunk 3]
```diff
@@ -73,7 +73,7 @@ const handleModelHover = async () => {
     modelPreviewStyle.value.left = `${targetRect.left - 400}px`
   }
 
-  modelDef.value.load()
+  modelDef.value.load?.()
```
Comment (huchenlei) [inline src/components/sidebar/tabs/modelLibrary/ModelTreeLeaf.vue]: nit: revert unnecessary change here.

File: src/components/common/ElectronFileDownload.vue [hunk 4]
```diff
@@ -81,55 +83,37 @@ const props = defineProps<{
...
-const downloads: ModelDownload[] = await DownloadManager.getAllDownloads()
-const modelDownload = downloads.find(({ url }) => url === props.url)
+electronDownloadStore.$subscribe((mutation, { downloads }) => {
+  const download = downloads.find((download) => props.url === download.url)
 
-const updateProperties = (download: ModelDownload) => {
-  if (download.url === props.url) {
-    status.value = download.status
+  if (download) {
     downloadProgress.value = (download.progress * 100).toFixed(1)
```
Comment (huchenlei) [inline src/components/common/ElectronFileDownload.vue]: Here we are assigning a string to a number. Can we move the number conversion here?

File: src/components/sidebar/tabs/ModelLibrarySidebarTab.vue [hunk 5]
```diff
@@ -28,6 +28,67 @@
       />
     </template>
     <template #body>
+      <div class="mx-6 mb-4" v-if="downloads.length > 0">
```
Comment (huchenlei) [inline src/components/sidebar/tabs/ModelLibrarySidebarTab.vue:31]: Let's extract this download section as a new component.

**Review by huchenlei — CHANGES_REQUESTED** [review]

**Review by huchenlei — APPROVED** [review]
LGTM! Thanks for addressing all comments.
